src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/UriPresentation/RemoteUriPresentationService.cs

```diff
@@ -0,0 +1,97 @@
+using Microsoft.ServiceHub.Framework;
+
+namespace Microsoft.CodeAnalysis.Remote.Razor;
+
+internal sealed class RemoteUriPresentationService(
+    IServiceBroker serviceBroker,
+    IRazorDocumentMappingService documentMappingService,
+    DocumentSnapshotFactory documentSnapshotFactory,
+    ILoggerFactory loggerFactory)
+    : RazorServiceBase(serviceBroker), IRemoteUriPresentationService
+{
+    private readonly IRazorDocumentMappingService _documentMappingService = documentMappingService;
+    private readonly DocumentSnapshotFactory _documentSnapshotFactory = documentSnapshotFactory;
+    private readonly ILogger _logger = loggerFactory.GetOrCreateLogger<RemoteUriPresentationService>();
+
+    public ValueTask<TextChange?> GetPresentationAsync(RazorPinnedSolutionInfoWrapper solutionInfo, DocumentId razorDocumentId, LinePositionSpan span, Uri[]? uris, CancellationToken cancellationToken)
+    => RazorBrokeredServiceImplementation.RunServiceAsync(
+        solutionInfo,
+        ServiceBrokerClient,
+        solution => GetPresentationAsync(solution, razorDocumentId, span, uris, cancellationToken),
+        cancellationToken);
+
+    private async ValueTask<TextChange?> GetPresentationAsync(Solution solution, DocumentId razorDocumentId, LinePositionSpan span, Uri[]? uris, CancellationToken cancellationToken)
+    {
+        var razorDocument = solution.GetAdditionalDocument(razorDocumentId);
+        if (razorDocument is null)
+        {
+            return null;
+        }
+
+        var snapshot = _documentSnapshotFactory.GetOrCreate(razorDocument);
+        var codeDocument = await snapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
+        var sourceText = await razorDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+        if (!sourceText.TryGetAbsoluteIndex(span.Start.Line, span.Start.Character, out var index))
+        {
+            return null;
+        }
+
+        var languageKind = _documentMappingService.GetLanguageKind(codeDocument, index, rightAssociative: true);
+
+        var razorFileUri = UriPresentationHelper.GetComponentFileNameFromUriPresentationRequest(languageKind, uris, _logger);
+        if (razorFileUri is null)
+        {
+            return null;
+        }
+
+        var ids = razorDocument.Project.Solution.GetDocumentIdsWithFilePath(GetDocumentFilePathFromUri(razorFileUri));
+        if (ids.Length == 0)
+        {
+            return null;
+        }
+
+        // We assume linked documents would produce the same component tag so just take the first
+        var otherDocument = razorDocument.Project.Solution.GetAdditionalDocument(ids.First());
+        if (otherDocument is null)
+        {
+            return null;
+        }
+
+        var otherSnapshot = _documentSnapshotFactory.GetOrCreate(otherDocument);
+        var descriptor = await otherSnapshot.TryGetTagHelperDescriptorAsync(cancellationToken).ConfigureAwait(false);
+
+        if (descriptor is null)
```

Comment:
>if (descriptor is null) [](http://example.com/codeflow?start=8&length=23)

Both these and null check below can legitimately happen if tag helper discovery isn't finished yet?

Reply:
Yes, it also other reasons. A razor file might not represent a component that is known to this project for a few different scenarios.